File: android/src/main/java/com/rngooglemapsplus/MapPolygonBuilder.kt
package com.rngooglemapsplus

import android.graphics.Color
import com.facebook.react.uimanager.PixelUtil.dpToPx
import com.google.android.gms.maps.model.Polygon
import com.google.android.gms.maps.model.PolygonOptions
import com.rngooglemapsplus.extensions.coordinatesEquals
import com.rngooglemapsplus.extensions.holesEquals
import com.rngooglemapsplus.extensions.onUi
import com.rngooglemapsplus.extensions.toColor
import com.rngooglemapsplus.extensions.toLatLng
import com.rngooglemapsplus.extensions.toMapsPolygonHoles

class MapPolygonBuilder {
  fun build(poly: RNPolygon): PolygonOptions =
    PolygonOptions().apply {
      poly.coordinates.forEach { pt ->
        add(
          pt.toLatLng(),
        )
      }
      poly.fillColor?.let { fillColor(it.toColor()) }
      poly.strokeColor?.let { strokeColor(it.toColor()) }
      poly.strokeWidth?.let { strokeWidth(it.dpToPx()) }
      poly.pressable?.let { clickable(it) }
      poly.geodesic?.let { geodesic(it) }
      poly.holes?.forEach { hole ->
        addHole(hole.coordinates.map { it.toLatLng() })
      }
      poly.zIndex?.let { zIndex(it.toFloat()) }
    }

  fun update(
    prev: RNPolygon,
    next: RNPolygon,
    poly: Polygon,
  ) = onUi {
    if (!prev.coordinatesEquals(next)) {
      poly.points = next.coordinates.map { it.toLatLng() }
    }

    if (!prev.holesEquals(next)) {
      poly.holes = next.holes?.toMapsPolygonHoles() ?: emptyList()
    }

    if (prev.fillColor != next.fillColor) {
      poly.fillColor = next.fillColor?.toColor() ?: Color.TRANSPARENT
    }

    if (prev.strokeColor != next.strokeColor) {
      poly.strokeColor = next.strokeColor?.toColor() ?: Color.BLACK
    }

    if (prev.strokeWidth != next.strokeWidth) {
      poly.strokeWidth = next.strokeWidth?.dpToPx() ?: 1f
    }

    if (prev.pressable != next.pressable) {
      poly.isClickable = next.pressable ?: false
    }

    if (prev.geodesic != next.geodesic) {
      poly.isGeodesic = next.geodesic ?: false
    }

    if (prev.zIndex != next.zIndex) {
      poly.zIndex = next.zIndex?.toFloat() ?: 0f
    }
  }
}

File: ios/MapPolygonBuilder.swift
import GoogleMaps

final class MapPolygonBuilder {
  func build(_ p: RNPolygon) -> GMSPolygon {
    let path = p.coordinates.toGMSPath()
    let pg = GMSPolygon(path: path)

    p.fillColor.map { pg.fillColor = $0.toUIColor() }
    p.strokeColor.map { pg.strokeColor = $0.toUIColor() }
    p.strokeWidth.map { pg.strokeWidth = CGFloat($0) }
    p.pressable.map { pg.isTappable = $0 }
    p.geodesic.map { pg.geodesic = $0 }
    pg.holes = p.holes.toMapPolygonHoles()
    p.zIndex.map { pg.zIndex = Int32($0) }

    return pg
  }

  func update(_ prev: RNPolygon, _ next: RNPolygon, _ pg: GMSPolygon) {
    onMain {
      if !prev.coordinatesEquals(next) {
        pg.path = next.coordinates.toGMSPath()
      }

      if !prev.holesEquals(next) {
        pg.holes = next.holes.toMapPolygonHoles()
      }

      if prev.fillColor != next.fillColor {
        pg.fillColor = next.fillColor?.toUIColor() ?? .clear
      }

      if prev.strokeColor != next.strokeColor {
        pg.strokeColor = next.strokeColor?.toUIColor() ?? .black
      }

      if prev.strokeWidth != next.strokeWidth {
        pg.strokeWidth = CGFloat(next.strokeWidth ?? 1.0)
      }

      if prev.pressable != next.pressable {
        pg.isTappable = next.pressable ?? false
      }

      if prev.geodesic != next.geodesic {
        pg.geodesic = next.geodesic ?? false
      }

      if prev.zIndex != next.zIndex {
        pg.zIndex = Int32(next.zIndex ?? 0)
      }
    }
  }
}
